runtime/lua/vim/iter.lua

```diff
@@ -1019,12 +1020,13 @@ end
 ---
 ---@see |Iter:map()|
 ---
+---@param src table|function Table or iterator function to filter
 ---@param f function(...):?any Map function. Accepts the current iterator or table values as
 ---                            arguments and returns one or more new values. Nil values are removed
 ---                            from the final table.
----@param src table|function Table or iterator function to filter
 ---@return table
-function M.map(f, src, ...)
+function M.map(src, f, ...)
+  assert(type(f) == 'function', 'arg2 must be the callback function. :help news-breaking-dev')
   return Iter.new(src, ...):map(f):totable()
```

Comment:
Why is there "magic" happening with `M.map` and `M.filter`? Why don't they act like normal, non-"self", non-object, module functions?

```
test/functional/helpers.lua:107: Error executing lua: /home/runner/work/neovim/neovim/runtime/lua/vim/iter.lua:955: bad argument #1 to 'src' (table expected, got nil)
stack traceback:
	[C]: in function 'src'
	/home/runner/work/neovim/neovim/runtime/lua/vim/iter.lua:955: in function 'next'
	/home/runner/work/neovim/neovim/runtime/lua/vim/iter.lua:236: in function 'next'
	/home/runner/work/neovim/neovim/runtime/lua/vim/iter.lua:315: in function 'map'
	/home/runner/work/neovim/neovim/runtime/lua/vim/snippet.lua:426: in function 'append_to_snippet'
	/home/runner/work/neovim/neovim/runtime/lua/vim/snippet.lua:472: in function 'expand'
	[string "<nvim>"]:1: in main chunk

stack traceback:
	test/functional/helpers.lua:107: in function 'exec_lua'
	test/functional/lua/snippet_spec.lua:34: in function 'test_expand_success'
	test/functional/lua/snippet_spec.lua:45: in function <test/functional/lua/snippet_spec.lua:44>
```

Reply:
The entire vim.iter API is designed to work with any kind of Lua iterator. Iterators in Lua are simply functions that return multiple values. The iterator must be the final argument in these functions for this reason.

This means you can do, eg

```
vim.iter.map(cb, string.gmatch(...))
```

or

```
vim.iter.filter(cb, vim.fs.parents(...))
```

If you make the callback the last function argument then we lose the ability to use any function iterator with this API

Reply:
Yeah I keep forgetting this behavior of multiple return values. I was hoping a signature like `M.map(...)` and using `select()` to inspect the `...` would work. But that doesn't work because if one calls `map(ipairs(), fn)`, Lua drops but all but the first return value from `ipairs()`.

Reply:
Other parts of the stdlib that accept iterators could require either a table, function or `vim.iter`. Then they don't need to worry about placing the parameter at the end.

That would be another reason why avoiding the "scan" would be good: https://github.com/neovim/neovim/issues/24746

Reply:
We're also (and primarily) talking about the _Lua_ stdlib, though, which is not under our control...

Reply:
We don't control the function signatures of the Lua stdlib so I'm not worried about that :)

Reply:
But the point is that we want to chain _our_ iterators with Lua standard iterators (after all, chaining is their whole point), so we are constrained by _their_ function signatures.

Reply:
Example? If our stdlib functions accept function/table/vim.iter, the caller can convert those to Lua-stdlib arguments on-demand. But the Nvim-stdlib functions only need to _accept_ function/table/vim.iter, they don't need to _accept_ the multiple-return-value (ipairs-like) iterator.

Reply:
> they don't need to accept the multiple-return-value (ipairs-like) iterator.

That would severely limit the usefulness of this API; see @gpanders comment above.

Reply:
No. `vim.iter()` can accept multi-return iterator, but there's no need for every Nvim-stdlib function to also accept multi-return iterators. And it's impossible for things like zip() that accept multiple iterators, anyway.

This was also discussed in April: https://github.com/neovim/neovim/pull/23029#issuecomment-1517187405

Reply:
> But of course `vim.iter()` would still have to accept the multi-return-value type iterators.

Yes, absolutely. The key is that `vim.iter` is the common unit that is passed around, then we don't need to fashion every other Nvim-stdlib functon to directly handle mutli-return iterators.